src/System.Private.ServiceModel/src/System/ServiceModel/Security/SendSecurityHeader.cs

```diff
@@ -40,9 +130,853 @@ public override string Namespace
+                return;
+            }
+
+            if (_skipKeyInfoForEncryption)
+            {
+                WrappedKeySecurityToken wrappedKeyToken = _encryptingToken as WrappedKeySecurityToken;
+                wrappedKeyToken.EnsureEncryptedKeySetUp();
+                wrappedKeyToken.EncryptedKey.ReferenceList = (ReferenceList)referenceList;
+            }
+            else
+            {
+                _elementContainer.ReferenceList = referenceList;
+            }
+            _basicTokenEncrypted = true;
+        }
+
+        internal void StartSecurityApplication()
+        {
+            if (this.SignThenEncrypt)
+            {
+                StartSignature();
+                StartEncryption();
+            }
+            else
+            {
+                StartEncryption();
+                StartSignature();
+            }
+        }
+
+        internal void CompleteSecurityApplication()
+        {
+            if (this.SignThenEncrypt)
+            {
+                CompleteSignature();
+                SignWithSupportingTokens();
+                CompleteEncryption();
+            }
+            else
+            {
+                CompleteEncryption();
+                CompleteSignature();
+                SignWithSupportingTokens();
+            }
+
+            if (_correlationState != null)
+            {
+                _correlationState.SignatureConfirmations = GetSignatureValues();
+            }
+        }
+
+        void AddGeneratedSignatureValue(byte[] signatureValue, bool wasEncrypted)
+        {
+            // cache outgoing signatures only on the client side
+            if (this.MaintainSignatureConfirmationState && (_signatureConfirmationsToSend == null))
+            {
+                if (_signatureValuesGenerated == null)
+                {
+                    _signatureValuesGenerated = new SignatureConfirmations();
+                }
+                _signatureValuesGenerated.AddConfirmation(signatureValue, wasEncrypted);
+            }
+        }
+    }
+
+    class TokenElement : ISecurityElement
+    {
+        SecurityStandardsManager standardsManager;
+        SecurityToken token;
+
+        public TokenElement(SecurityToken token, SecurityStandardsManager standardsManager)
+        {
+            this.token = token;
+            this.standardsManager = standardsManager;
+        }
+
+        public override bool Equals(object item)
+        {
+            TokenElement element = item as TokenElement;
+            return (element != null && this.token == element.token && this.standardsManager == element.standardsManager);
+        }
+
+        public override int GetHashCode()
+        {
+            return token.GetHashCode() ^ standardsManager.GetHashCode();
+        }
+
+        public bool HasId
+        {
+            get { return true; }
+        }
+
+        public string Id
+        {
+            get { return token.Id; }
```

Comment:
null ref possible - maybe we want to Contract.Assert at the ctor as a preemptive measure?